### ISSUE TITLE
Failed Slurm jobs result into pods with status "completed"

### ISSUE THREAD

**Issue by mbarbetti** [interTwin-eu/interlink-slurm-plugin#8]
# Short Description of the issue

I have just started to play with this Slurm plugin to contribute to letting it pass all the requirements of the [Virtual Kubelet Test Set](https://github.com/landerlini/vk-test-set). Even if the [`hello-world`](https://github.com/landerlini/vk-test-set/blob/main/vktestset/templates/000-hello-world.yaml) test fails, the corresponding pod returns with status completed. This issue motivates the need for an additional test to prevent similar problems, as described in https://github.com/landerlini/vk-test-set/issues/2.

## Environment

Two VMs hosted at Cloud@CNAF:
  1. `vm-source`
      * Ubuntu 24.04
      *  Running a k8s cluster with a virtual node based on [landerlini/interlink-virtual-node](https://github.com/landerlini/interlink-virtual-node)
  2. `vm-target`
      * Ubuntu 24.04
      * Running a Slurm cluster based on the docker-compose described by [Rodrigo Ancavil's Medium post](https://medium.com/analytics-vidhya/slurm-cluster-with-docker-9f242deee601)
      * The docker-compose has been updated to include an interLink server based on the [release 0.2.3-rc1](https://github.com/interTwin-eu/interLink/releases/tag/0.2.3-rc1)

## Steps to reproduce

1. Download the [Virtual Kubelet Test Set](https://github.com/landerlini/vk-test-set) in the machine that hosts the k8s cluster (`vm-source` in my case)
2. Move to the downloaded repository and configure the tests by editing the [`vktest_config.yaml`](https://github.com/landerlini/vk-test-set/blob/main/vktest_config.yaml) file
3. Install [pytest](https://pypi.org/project/pytest) and run all the tests enforcing stop after the first failure (use `pytest -vx`)
4. Go to the machine that hosts the Slurm cluster (`vm-target` in my case)
5. Check the logs of the container that runs the Slurm plugin (`slurm-virtual-node` in my case)

## Logs, stacktrace, or other symptoms

Output obtained by running `pytest -vx` from **`vm-source`**:

```
## Hello world
##
## First simple test to ensure the communication is open and a simple
## echo command can be executed.
## Launches a pod printing "hello world" in the log.

apiVersion: v1
kind: Pod
metadata:
  name: hello-world-d6726b45-e1d2-4838-9e55-240a4e111400
  namespace: interlink
  annotations: {"slurm-job.knoc.io/flags": "--job-name=test-pod-cfg -t 2800  --ntasks=8 --nodes=1 --mem-per-cpu=2000"}

spec:
  restartPolicy: Never
  nodeSelector:
    kubernetes.io/hostname: vm-target

  containers:
  - name: hello-world
    image: ghcr.io/grycap/cowsay 
    command: ["/bin/bash", "-c"]
    args: ["echo hello world"]
    imagePullPolicy: Always
    
  dnsPolicy: ClusterFirst

  tolerations: [{"effect": "NoSchedule", "key": "virtual-node.interlink/no-schedule", "operator": "Exists"}]

################################################################################
timeout_seconds=20.0 clean_configs=[CleanConfig(type='pod', name='hello-world-d6726b45-e1d2-4838-9e55-240a4e111400', namespace='interlink', condition='never')] check_pods=[CheckPod(name='hello-world-d6726b45-e1d2-4838-9e55-240a4e111400', namespace='interlink', status='Succeeded')] check_logs=[CheckLogs(name='hello-world-d6726b45-e1d2-4838-9e55-240a4e111400', namespace='interlink', regex='hello world', operator='Exists', container='hello-world', value=None)]
[]
[]
[]
[]
[]
[]
[]
[]
[]
[]
[]
[]
[]
[]
```

Log shown by running `docker logs slurm-virtual-node -f` from **`vm-target`**:

```
INFO[0000] Loading Slurm config from /etc/interlink/SlurmConfig.yaml
DEBU[0000] Debug level: true
ERRO[0000] Matteo's updates start here
DEBU[0000] open .local/interlink/jobs/interlink-3c6ca2c3-2f61-426c-bcc1-cec0dac7f008/StartedAt.time: no such file or directory
DEBU[0000] open .local/interlink/jobs/-/StartedAt.time: no such file or directory
DEBU[0000] open .local/interlink/jobs/-/FinishedAt.time: no such file or directory
INFO[0007] Slurm Plugin: received Submit call           
INFO[0007] - Beginning script generation for container hello-world 
INFO[0007] -- Preparing mountpoints for hello-world     
INFO[0007] -- Created directory .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838 
DEBU[0007] []                                           
DEBU[0007] -- Appending all commands together...        
INFO[0007] -- Creating file for the Slurm script        
INFO[0007] -- Created directory .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838 
DEBU[0007] --- Created file .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/job.sh 
DEBU[0007] --- Writing file                             
DEBU[0007] ---- Written file                            
INFO[0007] - Submitting Slurm job                       
DEBU[0007] Job submitted                                
INFO[0007] Submitted batch job 10                       
INFO[0007] Job ID is: 10                                
INFO[0011] Slurm Plugin: received GetStatus call        
INFO[0011] JID: 10 | Status:  | Pod: hello-world-d6726b45-e1d2-4838-9e55-240a4e111400 | UID: a78a82f5-c2be-4704-98e9-a88b4a1e2838 
DEBU[0011] [{hello-world-d6726b45-e1d2-4838-9e55-240a4e111400 a78a82f5-c2be-4704-98e9-a88b4a1e2838 interlink [{hello-world {nil nil &ContainerStateTerminated{ExitCode:0,Signal:0,Reason:,Message:,StartedAt:0001-01-01 00:00:00 +0000 UTC,FinishedAt:2024-05-31 16:17:25.542741974 +0000 UTC m=+11.928273599,ContainerID:,}} {nil nil nil} false 0    <nil> map[] nil []}]}] 
INFO[0012] Slurm Plugin: received GetLogs call          
INFO[0012] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0013] Slurm Plugin: received GetLogs call          
INFO[0013] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0015] Slurm Plugin: received GetLogs call          
INFO[0015] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0016] Slurm Plugin: received GetLogs call          
INFO[0016] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0017] Slurm Plugin: received GetLogs call          
INFO[0017] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0018] Slurm Plugin: received GetLogs call          
INFO[0018] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0019] Slurm Plugin: received GetLogs call          
INFO[0019] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0020] Slurm Plugin: received GetLogs call          
INFO[0020] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0021] Slurm Plugin: received GetLogs call          
INFO[0021] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0022] Slurm Plugin: received GetLogs call          
INFO[0022] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0023] Slurm Plugin: received GetLogs call          
INFO[0023] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0024] Slurm Plugin: received GetLogs call          
INFO[0024] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0025] Slurm Plugin: received GetLogs call          
INFO[0025] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out 
INFO[0026] Slurm Plugin: received GetLogs call          
INFO[0026] Reading  .local/interlink/jobs/interlink-a78a82f5-c2be-4704-98e9-a88b4a1e2838/hello-world.out
```

The reason why the Slurm job has failed:

```
/var/spool/slurmd/job00010/slurm_script: line 6: singularity: command not found
```

# Summary of proposed changes

Trying to fix this issue and aiming to contribute to developing this Slurm plugin, I have [forked this repo](https://github.com/mbarbetti/interlink-slurm-plugin). I plan to keep this issue updated on the adopted solution and any further fixes. I expect to prepare a PR at some point.

**Comment by Surax98** [comment]
@mbarbetti so will you work on this and will you make a PR?

**Comment by Bianco95** [comment]
I have also running the tests to understand the behaviour of the plugin and I have also encountered the error on the hello-world test. There are two main problems: the first one is related to name of the image of the pod, indeed when I check the logs of the POD running the test I see this `FATAL:   While checking container encryption: could not open image /root/ghcr.io/grycap/cowsay:`. This problem can be temporally fixed using the root image name as `docker://`. After temporally fixed this problem, the second one appeared and it is related to the args that are not parsed correctly `args: ["echo hello world"]` `hello world` is not printed, so there is a problem with parsing the args.

**Comment by Surax98** [comment]
> I have also running the tests to understand the behaviour of the plugin and I have also encountered the error on the hello-world test. There are two main problems: the first one is related to name of the image of the pod, indeed when I check the logs of the POD running the test I see this `FATAL: While checking container encryption: could not open image /root/ghcr.io/grycap/cowsay:`. This problem can be temporally fixed using the root image name as `docker://`. After temporally fixed this problem, the second one appeared and it is related to the args that are not parsed correctly `args: ["echo hello world"]` `hello world` is not printed, so there is a problem with parsing the args.

Addressed in the [0.2.5-1](https://github.com/interTwin-eu/interlink-slurm-plugin/releases/tag/0.2.5-1) release.
Commit [ab203ca](https://github.com/interTwin-eu/interlink-slurm-plugin/commit/ab203ca2d6148d813c849f5101c1f7d9ab2b3d75)
It will be merged in the 0.3 later.

**Comment by Surax98** [comment]
Issue can be closed since error codes are now properly returned (commit [79c44fb](https://github.com/interTwin-eu/interlink-slurm-plugin/commit/79c44fb48a79558f4c184f3894db805a3c93aa19)) and set VK side.
The Completed status must be addressed VK side, but it's not a problem since it's just a Reason for the Pod being terminated, it's not the actual Status, as you can easily check with a kubectl describe pod:
```bash
kubectl get pods -n vk

NAME    READY   STATUS      RESTARTS   AGE
hello   0/4     Completed   0          77s
```

```bash
kubectl describe pod hello -n vk

Name:             hello
Namespace:        vk
Priority:         0
Service Account:  default
Node:             test-vk/
Start Time:       Thu, 13 Jun 2024 10:46:29 +0000
Labels:           <none>
Annotations:      slurm-job.knoc.io/flags: --job-name=test-pod-cfg -t 2800  --ntasks=8 --nodes=1 --mem-per-cpu=2000
Status:           Failed
Reason:           Error: 
IP:               
IPs:              <none>
Containers:
  bash:
    Container ID:  
    Image:         docker://ghcr.io/grycap/cowsay
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/bash
      -c
    Args:
      echo hello world
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 13 Jun 2024 10:46:31 +0000
      Finished:     Thu, 13 Jun 2024 10:46:51 +0000
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  1Mi
    Requests:
      cpu:        1
      memory:     1Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jgfx6 (ro)
  python:
    Container ID:  
    Image:         docker://ghcr.io/grycap/cowsay
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      python
      -c
    Args:
      print('Hello, World!')
    State:          Terminated
      Reason:       Error: 
      Exit Code:    127
      Started:      Thu, 13 Jun 2024 10:46:31 +0000
      Finished:     Thu, 13 Jun 2024 10:46:51 +0000
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  1Mi
    Requests:
      cpu:        1
      memory:     1Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jgfx6 (ro)
  curl:
    Container ID:  
    Image:         docker://ghcr.io/grycap/cowsay
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      curl
      -X
    Args:
      POST localhost
    State:          Terminated
      Reason:       Error: 
      Exit Code:    127
      Started:      Thu, 13 Jun 2024 10:46:31 +0000
      Finished:     Thu, 13 Jun 2024 10:46:51 +0000
    Ready:          False
    Restart Count:  0
    Limits:
      memory:  1Mi
    Requests:
      memory:     1Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jgfx6 (ro)
  piped:
    Container ID:  
    Image:         docker://ghcr.io/grycap/cowsay
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      curl
      -X
    Args:
      POST localhost && echo henlo
    State:          Terminated
      Reason:       Error: 
      Exit Code:    127
      Started:      Thu, 13 Jun 2024 10:46:31 +0000
      Finished:     Thu, 13 Jun 2024 10:46:51 +0000
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:  1
    Requests:
      cpu:        1
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jgfx6 (ro)
  ```
Specifically, for reasons which are still under investigations, this Reason is set to the Reason of the first Container in the list, so it can be both Completed or an error depending on the Exit Code of the first Container. It is not very problematic, luckily, since it's just a "at a glance" answer to the question "why did it terminate?", but the Pod Status is actually set to Failed.
